### PR TITLE
bunch of fixes (overcast, hidden opposed tests, opposed test NaN for unopposed)

### PR DESCRIPTION
**Pull request by silentmark**
Hi, 

bunch of fixes. 
1. whenever an overcast is used for duration - it is not applied for active effect, because test is recreated from data inside message. 
2. since mask flag is not used anymore - i fallback into hidden tokens, so now if hidden token will start opposed test - his name/icon will not be revealed. 
3. overcast for player being applicable on failed test
4. when opposed test was changed to unopposed, sometimes NaN value was showing up for me.
5. active effects now will be run in async or normal mode depending on the script.
6. there is still some issues with 'other' information in tests.

### CODE REVIEW

File: modules/actor/actor-wfrp4e.js [hunk 1]
```diff
@@ -3606,19 +3584,23 @@ export default class ActorWfrp4e extends Actor {
...
+    if (duration && item.duration.value.toLowerCase().includes(game.i18n.localize("Minutes")))
+      effect.duration.seconds = duration * 60 * multiplier
 
-    else if (item.duration && item.duration.value.toLowerCase().includes(game.i18n.localize("hours")))
-      effect.duration.seconds = parseInt(item.Duration) * 60 * 60 * multiplier
+    else if (duration && item.duration.value.toLowerCase().includes(game.i18n.localize("Hours")))
+      effect.duration.seconds = duration * 60 * 60 
 
-    else if (item.duration && item.duration.value.toLowerCase().includes(game.i18n.localize("rounds")))
-      effect.duration.rounds = parseInt(item.Duration) * multiplier
```
Comment (moo-man) [inline modules/actor/actor-wfrp4e.js:3621]: > whenever an overcast is used for duration - it is not applied for active effect, because test is recreated from data inside message.

I can't seem to reproduce this. The test (and message) data should be updated with every overcast and calculate the correct value in `usage.duration`

Additionally your fix uses the same object, so why would that be different? (`duration.count` vs `duration.current`

File: modules/hooks/combat.js [hunk 2]
```diff
@@ -7,12 +7,11 @@ export default function() {
   Hooks.on("deleteCombat", CombatHelpers.endCombat)
 
   Hooks.on("preCreateCombatant", (combatant, data) => {
-    let mask = canvas.tokens.get(data.tokenId).document.getFlag("wfrp4e", "mask")
+    let mask = canvas.tokens.get(data.tokenId).document.hidden;
```
Comment (moo-man) [inline modules/hooks/combat.js:10]: This should use `combatant.hidden`

File: modules/hooks/handlebars.js [hunk 3]
```diff
@@ -27,11 +27,13 @@ export default function () {
...
 
         Handlebars.registerHelper("tokenImg", function(actor) {
-            return actor.token ? actor.token.texture.src : actor.prototypeToken.texture.src
+            let token = game.canvas.tokens.getDocuments().filter(x=>x.actorId == actor.id)[0] ?? actor.prototypeToken;
+            return token.hidden ? "systems/wfrp4e/tokens/unknown.png" : token.texture.src;
         })
 
         Handlebars.registerHelper("tokenName", function(actor) {
-            return actor.token ? actor.token.name : actor.prototypeToken.name;
+            let token = game.canvas.tokens.getDocuments().filter(x=>x.actorId == actor.id)[0] ?? actor.prototypeToken;
```
Comment (moo-man) [inline modules/hooks/handlebars.js:35]: This is really gross but I think to make it cleaner you could just use `actor.getActiveTokens()`

File: modules/system/combat.js [hunk 4]
```diff
@@ -56,8 +56,12 @@ export default class CombatHelpers {
             if (turn.actor.hasSystemEffect("dualwielder"))
                 turn.actor.removeSystemEffect("dualwielder")
 
-            if (game.settings.get("wfrp4e", "statusOnTurnStart"))
-                turn.actor.displayStatus(combat.round, turn.name);
+            if (game.settings.get("wfrp4e", "statusOnTurnStart")) {
+                let token = game.canvas.tokens.get(combat.current.tokenId);
+                let nameOverride = token.document.hidden ? "???" : turn.name;
+
+                turn.actor.displayStatus(combat.round, nameOverride);
```
Comment (moo-man) [inline modules/system/combat.js:63]: Can use `game.combat.combatant` here

File: modules/system/opposed-test.js [hunk 5]
```diff
@@ -161,8 +161,8 @@ export default class OpposedTest {
 
       opposeResult.other = opposeResult.other.concat(opposeResult.modifiers.message);
 
-      let attackerSL = parseInt(attackerTest.result.SL);
-      let defenderSL = parseInt(defenderTest.result.SL);
+      let attackerSL = parseInt(attackerTest.result.SL ?? 0);
```
Comment (moo-man) [inline modules/system/opposed-test.js:164]: > when opposed test was changed to unopposed, sometimes NaN value was showing up for me.

Was this the fix? What do you mean by opposed to unopposed? Are there reproduction steps?

File: modules/system/opposed-wfrp4e.js [hunk 6]
```diff
@@ -115,9 +115,14 @@ export default class OpposedWFRP {
 
   renderOpposedStart() {
     return new Promise(async resolve => {
-      let attacker = WFRP_Utility.getToken(this.attackerTest.context.speaker) || this.attacker.prototypeToken;
+      let attacker = game.canvas.tokens.get(this.attackerTest.context.cardOptions.speaker.token)?.document ?? this.attacker.prototypeToken;
       let defender
```
Comment (moo-man) [inline modules/system/opposed-wfrp4e.js:119]: why wasn't `getToken` working here?

File: modules/system/rolls/test-wfrp4e.js [hunk 7]
```diff
@@ -535,7 +535,13 @@ export default class TestWFRP {
...
 
     //this.result.other = this.result.other.join("<br>")
-    this.result.other = this.preData.other.join("<br>")
+
+    if (this.preData.other.length > 0) {
+      this.result.other = this.preData.other.join("<br>");
+    }
+    else if(this.result.other && this.result.other.join) {
+      this.result.other = this.result.other.join("<br>");
+    }
```
Comment (moo-man) [inline modules/system/rolls/test-wfrp4e.js:544]: Not sure what these two cases are for? Couldn't it just be
```js
if (this.result.other instanceof Array)
{
   this.result.other = this.result.other.join("<br>")
}
```

File: modules/system/utility-wfrp4e.js [hunk 8]
```diff
@@ -1080,9 +1080,30 @@ export default class WFRP_Utility {
...
+      }
+      if(func) {
+        func(scriptArgs);
+      }
+    }
+    catch (ex) {
+      ui.notifications.error(game.i18n.format("ERROR.EFFECT", { effect: effect.label }))
+      console.error("Error when running effect " + effect.label + " - If this effect comes from an official module, try replacing the actor/item from the one in the compendium. If it still throws this error, please use the Bug Reporter and paste the details below, as well as selecting which module and 'Effect Report' as the label.")
+      console.error(`REPORT\n-------------------\nEFFECT:\t${effect.label}\nACTOR:\t${actor.name} - ${actor.id}\nERROR:\t${ex}`)
+    }
```
Comment (moo-man) [inline modules/system/utility-wfrp4e.js:1106]: Why was this needed exactly? There are specific triggers that can't support asynchronous execution and specific ones that can. The specific ones were given async functions already.

File: modules/actor/actor-wfrp4e.js (hunk 1 above)
Comment (silentmark) [reply modules/actor/actor-wfrp4e.js:3634]: so, the change here came during my testing and was purely cosmetical. The only potential benefit of this was to use exact duration in case there are any other homebrew in use. Anyway, here is example where problem lies: 

image

The roll is recreated from message flag after overcast click, but the flags are not updated, thats why i have added the:

      this.updateMessageFlags();

into all overcast and overcastreset messages.

File: modules/system/opposed-test.js (hunk 5 above)
Comment (silentmark) [reply modules/system/opposed-test.js:164]: Ok, so now i have problems to reproduce this as well, but the issue which i've been facing was related to result.SL being undefined and parseInt was returning me NaN.

File: modules/system/opposed-wfrp4e.js (hunk 6 above)
Comment (silentmark) [reply modules/system/opposed-wfrp4e.js:119]: image


Scene was null for speaker so it was always falling back to token prototype

File: modules/system/rolls/test-wfrp4e.js (hunk 7 above)
Comment (silentmark) [reply modules/system/rolls/test-wfrp4e.js:544]: Sorry for that, it was quick and dirty during session, so i forgot to fix it before commit.

File: modules/system/utility-wfrp4e.js (hunk 8 above)
Comment (silentmark) [reply modules/system/utility-wfrp4e.js:1118]: first of all, to unify the logging errors and how effects are executed. Second of all, to allow async/await in all effects, because its not documented anywhere why and which effects may be async and which not. And even if - i dont get why only certain effects can be async? you are not awaiting any method in runEffects so it doesnt matter. In fact, you might want to look into my experimental branch https://github.com/silentmark/WFRP4e-FoundryVTT/tree/features/experimental where im trying to await all functiosn that are async, including effects.

Comment (moo-man) [reply modules/system/utility-wfrp4e.js:1118]: Any effect that gets called during actor preparation for example. Or really any time the calling function isn't async, calling effects that are won't really work / will be unpredictable.

Comment (moo-man) [reply modules/system/utility-wfrp4e.js:1118]: While I quite like unifying the code here, something about searching for "await" rubs me the wrong way, can't we still just use the `async` flag?

Comment (moo-man) [reply modules/system/utility-wfrp4e.js:1118]: If you look at the current implementation, `async` is just an option passed
```js
        if (!options.async)
          func = new Function("args", e.script).bind({ actor: this, effect: e, item: e.item })
        else if (options.async) {
          let asyncFunction = Object.getPrototypeOf(async function () { }).constructor
          func = new asyncFunction("args", e.script).bind({ actor: this, effect: e, item: e.item })
        }
```

It is underutilized however, I generally add it to a trigger when I need it to work for a specific effect and if the trigger supports it. Using callbacks was from before I added async support to effect scripts at all and tend to not do that anymore, only when I'm copy-pasting or it requires it.

`turn.actor.runEffects("endRound", combat, {async: true})`

As for what supports them, what absolutely cannot are preparation ones, such as `(pre)prepareData`, `(pre)prepareItem(s)`, `preWoundCalc`, etc. A comprehensive list would be a good thing to add to a hypothetical effects tutorial

As for why, an example would be an async `preWoundCalc` changes `sb`, `tb`, or any other of its arguments, won't happen in time for the calculation.

What could possibly support them would be `prefillDialog`? The calling function isn't async, so I think it would be unpredictable, and should probably change it to `async` first.

Point is, the `async` flag should be used and only on the ones that can fully support it. 

I would ask to hold off changing existing triggers (like `prefillDialog`) to support async for a separate PR though

File: modules/system/rolls/test-wfrp4e.js (hunk 7 above)
Comment (moo-man) [reply modules/system/rolls/test-wfrp4e.js]: Now it still has these two cases (if/else)

```js
    if (this.preData.other.length > 0) {
      this.result.other = this.preData.other.join("<br>");
    } 
    else if (this.result.other instanceof Array) {
       this.result.other = this.result.other.join("<br>")
    }
```

But as my previous comment suggests, can't it be one if statement testing if it's an array?

File: modules/actor/actor-wfrp4e.js (hunk 1 above)
Comment (moo-man) [reply modules/actor/actor-wfrp4e.js:3634]: Do you know of a case where a homebrew duration would fail?

Additionally
```js
 if (duration && item.duration.value.toLowerCase().includes(game.i18n.localize("Minutes")))
 ``` 
Why were the units of time capitalized? I'm not sure how it could be working because lowercase duration can never match "Minutes"

Lastly, 
If the `multiplier` variable is not being used it should be removed.

File: modules/system/utility-wfrp4e.js (hunk 8 above)
Comment (silentmark) [reply modules/system/utility-wfrp4e.js:1118]: sure, i will change it, but i will leave the unified function. Once i will have more time to test my experimental branch, i will create a PR that should allow to use async functions in more broad way, and also most of them should be properly awaited. I think that this lack of awaiting asynchronous functions sometimes leads to those anoying errors with CreateEmbeddedDocument or DeleteEmbeddedDocument.

File: modules/actor/actor-wfrp4e.js (hunk 1 above)
Comment (silentmark) [reply modules/actor/actor-wfrp4e.js]: Removed, also thats a funny thing - in en.json and other translations, Minutes, Rounds etc are all translated to lower case. IT was a small bummer that i found as well.

File: modules/system/chat-wfrp4e.js [hunk 9]
```diff
@@ -414,7 +418,7 @@ export default class ChatWFRP {
     let message = game.messages.get(msgId)
     let conditionResult;
 
-    if (combatant.actor.isOwner)
+    if (combatant.actor.isOwner && game.user.isGM)
```
Comment (moo-man) [inline modules/system/chat-wfrp4e.js:421]: Why was this added? Players applying their own conditions is preferable

File: modules/system/rolls/test-wfrp4e.js (hunk 7 above)
Comment (moo-man) [reply modules/system/rolls/test-wfrp4e.js]: Bumping this as I'd like to merge soon

File: modules/system/chat-wfrp4e.js (hunk 9 above)
Comment (silentmark) [reply modules/system/chat-wfrp4e.js:421]: crap, sorry, thanks for spotting this, it was for my personal use.

File: modules/system/rolls/test-wfrp4e.js (hunk 7 above)
Comment (silentmark) [reply modules/system/rolls/test-wfrp4e.js]: i would have to try recreate the issue one more time, but i noticed again, that in some cases, the preData.other is empty, and result.other is already an Array and not string. thats why two statements. it is something that "works for me"

Comment (moo-man) [reply modules/system/rolls/test-wfrp4e.js]: Well, as it is, it will throw an error if `this.preData.other` is a string, because `join` is not a string function. 

If we need two cases though, I think it's simpler and cleaner to combine them using `&&` conditional

Comment (silentmark) [reply modules/system/rolls/test-wfrp4e.js]: Ok, i've made another commit with proposal to get rid of preData.other as it was introduced purely for unofficialgirmoire. I hope this could solve the problem in clean way.

Comment (moo-man) [reply modules/system/rolls/test-wfrp4e.js]: Well this PR is already way out of scope for just being fixes, can you remove all changes related to `other` and put that into another PR?

Comment (silentmark) [reply modules/system/rolls/test-wfrp4e.js]: ok, changes are reverted and separate PR has been created for that
